Use cilium-based library for eBPF operations in network tracer

brycekahle opened the pull request:
### What does this PR do?

This switches to the cilium-based forked library at https://github.com/DataDog/ebpf 

- Uses runtime-replaced constants where possible to improve performance. Offset guessing is done on start (via a separate `C` file and manager), and those values are constants now instead of using a map for lookup on every kprobe hit.
- Uses tracepoints for syscalls (if enabled via config, default `false`), falling back to x64 compatible kprobe points, and finally the backwards-compatible kprobe points.
- Cleans up the system_probe build task to make it easier to add future targets.

### Motivation

This library is the way forward for the networks and security team's usage of eBPF. It offers improved performance, due to its reduction in `cgo` operations.

### Additional Notes

Waiting on the following PRs:
- ~~https://github.com/DataDog/ebpf/pull/2~~
- ~~https://github.com/DataDog/ebpf/pull/3~~
- ~~https://github.com/DataDog/ebpf/pull/4~~

### Describe your test plan

- Automated tests all pass
- Performance checked on load testing cluster

brycekahle committed:
Use cilium eBPF library for polling the perf maps

brycekahle committed:
Determine num of CPUs at runtime to prevent issues with gobpf/cilium loading the perf buffer  map

brycekahle committed:
duplicate the perf map FD before handing off to cilium to prevent Close problems

brycekahle committed:
WIP

brycekahle committed:
WIP2

brycekahle committed:
Fix TracerExpvar test config

brycekahle committed:
WIP3

brycekahle committed:
Add indirect syscall support

brycekahle committed:
Use constant for turning on dns stats collection

brycekahle committed:
Update system-probe in response to ebpf lib changes

brycekahle committed:
Update ebpf library to trunk version

brycekahle committed:
Working tracepoints and x64 fallback

brycekahle committed:
Add config option to turn on tracepoints

brycekahle committed:
Improve build task for ebpf programs

brycekahle committed:
Improve config fallback from tracepoint to kprobes

brycekahle committed:
Run goimports

brycekahle committed:
Remove unused function

brycekahle committed:
Fix nettop

brycekahle committed:
Fix rebase issues

brycekahle committed:
Add tests for syscall choosing

brycekahle committed:
Fix expvar test

brycekahle committed:
Add missing config code for EnableTracepoints

brycekahle committed:
Fix licenses

brycekahle committed:
Add missing build tag to config_linux_test

brycekahle committed:
Reformat line to satisfy python linter

brycekahle committed:
Use syscall for latest timestamp, rather than eBPF map

brycekahle committed:
Update datadog/ebpf with utsname fix

brycekahle committed:
Update datadog/ebpf with syscall prefix search fix

brycekahle committed:
Fix NowNanoseconds for 32 bit architectures

leeavital commented on pkg/config/config.go:
```diff
@@ -676,6 +676,7 @@ func InitConfig(config Config) {
 	config.SetKnown("system_probe_config.offset_guess_threshold")
 	config.SetKnown("system_probe_config.enable_tcp_queue_length")
 	config.SetKnown("system_probe_config.enable_oom_kill")
+	config.SetKnown("system_probe_config.enable_tracepoints")
```
is there a performance gain to be gained from enabling tracepoints ?  We intercept so few syscalls, it doesn't feel like its worth supporting two code paths (for kprobes and tracepoints)

brycekahle replied:
There is a theoretical performance gain. They also have a stable interface, whereas kprobes can technically change in any kernel version (example: the x64 indirection we have to handle). See https://lwn.net/Articles/748352/ for perf comparison (along with raw tracepoints).

They share the same actual probe code, just the entry function is different, which passes along the arguments.

If we ever change our kernel targets so that tracepoints are always available, IMHO it would be preferable to use them.

sunhay replied:
I'm not opposed to adding select branches that offer increased performance for kernels after a particular version, if its available. I think this is a pattern we will (soon) be adopting widely as we support more use-cases via eBPF.

We'll definitely want to weigh the cost of supporting each additional path re: general testing and debugging of support cases. Since this is disabled by default, I think it's fine to keep it in while we further test before widely enabled based on detected kernel version.  

On the support side, we should probably add similar logging as bpftrace's `info` command to collect useful context:
- i.e https://github.com/iovisor/bpftrace/issues/1435#issue-664641877

brycekahle commented on pkg/ebpf/bytecode/manager.go:
```diff
@@ -0,0 +1,86 @@
+		Probes: []*manager.Probe{
+			{Section: string(TCPGetInfo)},
+			{Section: string(TCPv6Connect)},
+			{Section: string(TCPv6ConnectReturn), KProbeMaxActive: maxActive},
```
That option only applies to return kprobes

brycekahle committed:
Exempt time file from linter unconvert

brycekahle committed:
Move nolint comment because linter thinks it is a package comment

brycekahle committed:
Merge branch 'master' into bryce.kahle/cilium-ebpf

sunhay commented:
>Uses tracepoints for syscalls (if enabled via config, default false), falling back to x64 compatible kprobe points, and finally the backwards-compatible kprobe points.

Perhaps I'm missing something obvious, but what are we using the added tracepoints for currently? From a quick glance at the code - it seems like we're using it for determining UDP direction? Is that correct?

brycekahle commented:
> Perhaps I'm missing something obvious, but what are we using the added tracepoints for currently

It isn't new functionality, rather intercepting syscalls (`socket` and `bind`) that we were already probing. Those probes are indeed used for determining UDP direction.

brycekahle committed:
Limit time.go build to linux

p-lambert commented on pkg/ebpf/offsetguess.go:
```diff
@@ -380,8 +382,11 @@ func setReadyState(m *elf.Module, mp *elf.Map, status *tracerStatus) error {
 // guess the next field.
-func guessOffsets(m *elf.Module, cfg *Config) error {
-	mp := m.Map(string(tracerStatusMap))
+func guessOffsets(m *manager.Manager, cfg *Config) ([]manager.ConstantEditor, error) {
```
I think the comments need to be updated to reflect some of the changes

p-lambert commented on pkg/ebpf/tracer.go:
```diff
@@ -237,23 +229,72 @@ func NewTracer(config *Config) (*Tracer, error) {
+		return nil, fmt.Errorf("could not start offset ebpf manager: %s", err)
+	}
+	defer func() {
+		err := offsetMgr.Stop(manager.CleanAll)
```
Glad to see we're finally disabling the Kprobes used exclusively for offset guessing.

p-lambert commented on pkg/ebpf/c/tracer-ebpf.c:
```diff
@@ -379,20 +330,20 @@ static int read_conn_tuple(conn_tuple_t* t, tracer_status_t* status, struct sock
-        bpf_probe_read(&t->daddr_h, sizeof(t->daddr_h), ((char*)skp) + status->offset_daddr_ipv6);
-        bpf_probe_read(&t->daddr_l, sizeof(t->daddr_l), ((char*)skp) + status->offset_daddr_ipv6 + sizeof(u64));
+    } else if (is_ipv6_enabled() && check_family(skp, AF_INET6)) {
+        bpf_probe_read(&t->saddr_h, sizeof(t->saddr_h), ((char*)skp) + offset_daddr_ipv6() + 2 * sizeof(u64));
+        bpf_probe_read(&t->saddr_l, sizeof(t->saddr_l), ((char*)skp) + offset_daddr_ipv6() + 3 * sizeof(u64));
+        bpf_probe_read(&t->daddr_h, sizeof(t->daddr_h), ((char*)skp) + offset_daddr_ipv6());
+        bpf_probe_read(&t->daddr_l, sizeof(t->daddr_l), ((char*)skp) + offset_daddr_ipv6() + sizeof(u64));
```
This is great! Curious to see the perf impact of this

p-lambert commented on pkg/ebpf/c/tracer-ebpf.c:
```diff
@@ -561,104 +512,51 @@ static void cleanup_tcp_conn(struct pt_regs* ctx, conn_tuple_t* tup) {
     u64 pid_tgid = bpf_get_current_pid_tgid();
-    u64 zero = 0;
-
-    tracer_status_t* status = bpf_map_lookup_elem(&tracer_status, &zero);
-    if (status == NULL) {
-        return 0;
-    }
```
YAS

sunhay replied:
+1

p-lambert commented on pkg/ebpf/c/tracer-ebpf.c:
```diff
@@ -45,6 +46,8 @@
 #define log_debug(fmt, ...)
 #endif
 
+#define LOAD_CONSTANT(param, var) asm("%0 = " param " ll" : "=r"(var))
```
This a bit cryptic. Worth adding some comments!

Gui774ume replied:
This is the post-compilation constant edition feature of the new lib 🎉 

Essentially what this part does is insert a load instruction with an immediate value, into the variable provided in the output section of the `asm()` call. The inserted symbol is used so that the rewrite function is able to detect where the constant should be edited.
Have a look at https://github.com/DataDog/ebpf/blob/master/manager/editor.go#L26-L65.

p-lambert replied:
@Gui774ume very ingenious :)

p-lambert reviewed (commented):
Gave a first pass. Changes look really great!
Given how big this change is, I'd definitely deploy it in one of our internal clusters and test it thoroughly before merging to master.

brycekahle commented on pkg/ebpf/common_linux.go:
```diff
@@ -34,15 +34,6 @@ func isUbuntu(platform string) bool {
 	return strings.Contains(strings.ToLower(platform), "ubuntu")
 }
 
-func isCentOS(platform string) bool {
```
Yes

brycekahle committed:
Update comments of guessOffsets

brycekahle committed:
Add comment explaining LOAD_CONSTANT

brycekahle committed:
Update datadog/ebpf with android fix

brycekahle committed:
Merge branch 'master' into bryce.kahle/cilium-ebpf

brycekahle committed:
Add simple test for NowNanoseconds

brycekahle committed:
Cleanup headers

leeavital commented on pkg/ebpf/bytecode/manager.go:
```diff
@@ -0,0 +1,86 @@
+const (
+	// maxActive configures the maximum number of instances of the kretprobe-probed functions handled simultaneously.
+	// This value should be enough for typical workloads (e.g. some amount of processes blocked on the accept syscall).
+	maxActive = 128
```
Should this be `min(128, runtime.NumCPU())` ?

Gui774ume replied:
We've had multiple issues with `maxActive` on the security agent side. We'd recommend to keep a high value (our current `maxActive` is 512 for example) since `NumCPU` will not be enough for syscalls that can be pre-empted.

Also, after spending some (way too much) time on this, we also concluded that the second column in `kprobe_profile` is not the number of kretprobe misses (although we wouldn't be able to tell you what it is), and the only way to accurately keep track of the number of kretprobe misses is to calculate the difference between the number of kprobe and kretprobe hist for the same hook point.

brycekahle replied:
Note this wasn't a new value, just moved from `tracer.go`:
https://github.com/DataDog/datadog-agent/blob/bd0991d26ebcc5f02e3d25bad00bd96e292194fb/pkg/ebpf/tracer.go#L83-L85

leeavital commented on pkg/ebpf/c/asm_goto_workaround.h:
```diff
@@ -0,0 +1,26 @@
+#ifndef __ASM_GOTO_WORKAROUND_H
```
👍 glad to see this extracted into its own file

leeavital commented on pkg/ebpf/c/offset-guess.c:
```diff
@@ -0,0 +1,185 @@
+static int guess_offsets(tracer_status_t* status, struct sock* skp) {
+    u64 zero = 0;
+
+    if (status->state != TRACER_STATE_CHECKING) {
+        return 1;
+    }
```
I don't think this much matters, but do we still need this check now that this kprobe is removed once offset guessing is done?

brycekahle replied:
I think this is still required. Otherwise if the probe fires while userspace is figuring out if offset guessing is complete or not, it could do unnecessary/unwanted work.

Gui774ume commented:
I don't have enough context about how the tracer used to work to provide a valuable review, so I only had a look at the manager / new lib part and it looks great ! I'm excited to see our new lib in action 🎉

sunhay commented on pkg/network/network_filter.go:
```diff
@@ -149,9 +148,9 @@ func parsePortString(port string) (uint64, error) {
-// IsBlacklistedConnection returns true if a given connection should be excluded
+// IsExcludedConnection returns true if a given connection should be excluded
 // by the tracer based on user defined filters
-func IsBlacklistedConnection(scf []*ConnectionFilter, dcf []*ConnectionFilter, conn *ConnectionStats) bool {
```
🙇

brycekahle committed:
Run clang-format

sunhay commented on pkg/ebpf/bytecode/perf.go:
```diff
@@ -0,0 +1,27 @@
+	"github.com/DataDog/ebpf/manager"
+)
+
+type ClosedConnPerfHandler struct {
```
Nothing about this seems specific to closed conns re: types defined below. Should it just be called a `PerfHandler`?

brycekahle committed:
Rename to bytecode.PerfHandler

brycekahle committed:
Merge branch 'master' into bryce.kahle/cilium-ebpf

brycekahle committed:
Remove duplicate asm_volatile_goto workaround in security probe

brycekahle committed:
Fix license file

brycekahle committed:
Use intermediate bc file for system-probe build to get fast failure

brycekahle committed:
Merge branch 'master' into bryce.kahle/cilium-ebpf

brycekahle committed:
Merge branch 'master' into bryce.kahle/cilium-ebpf

brycekahle committed:
Ensure perf buffer goroutine exits

brycekahle committed:
Increase TCP batch timeout in tests so we don't hit max attempts before activity

brycekahle committed:
Use actual number of CPUs for number of batches

The MaxEntries of the tcp_close_event perf event array tells us the number of possible CPUs. See https://github.com/DataDog/ebpf/blob/d148b742bc922b7226f4e348e2d5723a1f5f3f58/map.go#L170

brycekahle committed:
Fix formatting on file

kevinconaway commented on pkg/ebpf/bytecode/perf.go:
```diff
@@ -0,0 +1,37 @@
+	}
+}
+
+func (c *PerfHandler) dataHandler(CPU int, batchData []byte, perfMap *manager.PerfMap, manager *manager.Manager) {
```
It looks like these calls are passed indirectly to `ebpf/manager`.  Is the shutdown process coordinated enough so that calls to the handler methods are guaranteed to have stopped before `Stop` is called and the channels are closed?

brycekahle replied:
Good question. The shutdown happens here: https://github.com/DataDog/datadog-agent/blob/7bc1ec2a457b8d49fc359e78dfdb417a49ed3285/pkg/ebpf/tracer.go#L404-L405

I can add some further coordination within the `PerfHandler` so it doesn't try to send to a closed channel.

brycekahle committed:
Ensure we don't attempt to send to closed channels in PerfHandler

p-lambert commented on pkg/ebpf/perf_batching.go:
```diff
@@ -116,23 +104,6 @@ func (p *PerfBatchManager) GetIdleConns(now time.Time) []network.ConnectionStats
 }
 
 type batchState struct {
-	offset   int
-	updated  int64
-	attempts int
-}
-
-// since we completely overshoot the number of CPU cores, we keep track of the number
-// of times we checked a batch and got not updates in order to detect which batches we can
-// ignore and avoid querying on eBPF.
-func (bs *batchState) ignore() bool {
-	if bs.attempts >= maxIgnoredCoreAttempts {
-		return true
-	}
-
-	// If there are no updates to this batch then the associated CPU core probably doesn't exist
-	if bs.updated == 0 {
-		bs.attempts++
-	}
-
-	return false
```
👌

brycekahle committed:
Merge branch 'master' into bryce.kahle/cilium-ebpf

brycekahle committed:
Add offset-guess to omnibus install

brycekahle committed:
Move TestProcessHeaders to run only when not bundling ebpf

hush-hush commented on .gitignore:
```diff
@@ -118,6 +118,9 @@ tools/windows/install-help/cal/packages/
 
 # ebpf object files
 pkg/ebpf/c/*.o
+pkg/ebpf/c/*.bc
+pkg/security/ebpf/c/*.o
+pkg/security/ebpf/c/*.bc
```
Should we just ignore `*.o` and `*.bc` files ?

brycekahle replied:
Probably 😄

brycekahle committed:
Merge branch 'master' into bryce.kahle/cilium-ebpf

brycekahle committed:
Merge branch 'master' into bryce.kahle/cilium-ebpf

brycekahle committed:
Merge branch 'master' into bryce.kahle/cilium-ebpf

brycekahle committed:
Add offset-guess to packaging

brycekahle committed:
Merge branch 'master' into bryce.kahle/cilium-ebpf

brycekahle committed:
Fix casing of expvar in tracer
